src/contracts/middleware/BLSPubkeyRegistry.sol

```diff
@@ -0,0 +1,208 @@
+     */
+    function getApkHashForQuorumAtBlockNumberFromIndex(uint8 quorumNumber, uint32 blockNumber, uint256 index) external returns (bytes32){
+        ApkUpdate memory quorumApkUpdate = quorumApkUpdates[quorumNumber][index];
+        _validateApkHashForQuorumAtBlockNumber(quorumApkUpdate, blockNumber);
+        return quorumApkUpdate.apkHash;
+    }
+
+	/**
+     * @notice get hash of the apk among all quourums at `blockNumber` using the provided `index`;
+     * called by checkSignatures in BLSSignatureChecker.sol.
+     */
+    function getGlobalApkHashAtBlockNumberFromIndex(uint32 blockNumber, uint256 index) external view returns (bytes32){
+        ApkUpdate memory globalApkUpdate = globalApkUpdates[index];
+        _validateApkHashForQuorumAtBlockNumber(globalApkUpdate, blockNumber);
+        return globalApkUpdate.apkHash;
+    }
+    /// @notice Returns the length of ApkUpdates for the provided `quorumNumber`
+    function getQuorumApkHistoryLength(uint8 quorumNumber) external view returns(uint32){
+        return uint32(quorumApkUpdates[quorumNumber].length);
+    }
+    /// @notice Returns the length of ApkUpdates for the global APK
+    function getGlobalApkHistoryLength() external view returns(uint32){
+        return uint32(globalApkUpdates.length);
+    }
+
+    function _processGlobalApkUpdate(BN254.G1Point memory point) internal {
+        // load and store in memory in common case we need to access the length again
+        uint256 globalApkUpdatesLength = globalApkUpdates.length;
+        // update the nextUpdateBlockNumber of the previous update
+        if (globalApkUpdates.length > 0) {
+            globalApkUpdates[globalApkUpdatesLength - 1].nextUpdateBlockNumber = uint32(block.number);
+        }
+
+        // accumulate the given point into the globalApk
+        globalApk = globalApk.plus(point);
+        // add this update to the list of globalApkUpdates
+        ApkUpdate memory latestGlobalApkUpdate;
+        latestGlobalApkUpdate.apkHash = BN254.hashG1Point(globalApk);
+        latestGlobalApkUpdate.updateBlockNumber = uint32(block.number);
+        globalApkUpdates.push(latestGlobalApkUpdate);
+    }
+
+    function _processQuorumApkUpdate(bytes memory quorumNumbers, BN254.G1Point memory point) internal {
+        BN254.G1Point memory apkBeforeUpdate;
+        BN254.G1Point memory apkAfterUpdate;
+
+        for (uint8 i = 0; i < uint8(quorumNumbers.length);) {
+            uint8 quorumNumber = uint8(quorumNumbers[i]);
+
+            uint256 quorumApkUpdatesLength = quorumApkUpdates[quorumNumber].length;
+            if (quorumApkUpdatesLength > 0) {
+                // update nextUpdateBlockNumber of the current latest ApkUpdate
+                quorumApkUpdates[quorumNumber][quorumApkUpdatesLength - 1].nextUpdateBlockNumber = uint32(block.number);
+            }
+            
+            apkBeforeUpdate = quorumApk[quorumNumber];
+
+            apkAfterUpdate = apkBeforeUpdate.plus(point);
```

Comment:
in this function this step is broken up into multiple lines, while in `_processGlobalApkUpdate`, this operation is combined into a single line.
Single line version seems more efficient (no unnecessary copying to memory) and IMO is at least as readable.